uportal-war/src/main/java/org/jasig/portal/groups/pags/testers/AdHocGroupTester.java

```diff
@@ -0,0 +1,208 @@
+import org.springframework.context.ApplicationContext;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.TreeSet;
+
+/**
+ * Immutable PAGS Tester for inclusive/exclusive membership in sets of groups. To avoid infinite recursion,
+ * calls to {@code test()} are tracked by parameters plus thread ID.
+ *
+ * @author  Benito J. Gonzalez <bgonzalez@unicon.net>
+ * @see     org.jasig.portal.groups.pags.IPersonTester
+ * @see     org.jasig.portal.groups.pags.dao.EntityPersonAttributesGroupStore
+ * @see     org.jasig.portal.groups.pags.dao.IPersonAttributesGroupTestGroupDefinition
+ * @since   4.3
+ */
+public final class AdHocGroupTester implements IPersonTester {
+
+    private final Cache currentTests;
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+    private final Set<IEntityGroup> includes = new HashSet<>();
+    private final Set<IEntityGroup> excludes = new HashSet<>();
+    private final String groupsHash;
+
+    public AdHocGroupTester(IPersonAttributesGroupTestDefinition definition) {
+        Validate.notNull(definition.getIncludes());
+        Validate.notNull(definition.getExcludes());
+        this.includes.addAll(collectGroups(definition.getIncludes(), true));
+        this.excludes.addAll(collectGroups(definition.getExcludes(), false));
+        this.groupsHash = calcGroupsHash(definition.getIncludes(), definition.getExcludes());
+        ApplicationContext context = ApplicationContextLocator.getApplicationContext();
+        CacheManager cacheManager = context.getBean("cacheManager", CacheManager.class);
+        this.currentTests = cacheManager.getCache("org.jasig.portal.groups.pags.testers.AdHocGroupTester");
+    }
+
+    /*
+     * At some point, a person is being tested for group membership. During that test, the thread hits an ad hoc group
+     * tester. When that tester calls isDeepMemberOf, a test for group membership is triggered. During this call stack,
+     * the second call the the ad hoc group tester returns false. Assuming the group hierarchy is not itself recursive
+     * for the group containing the ad hoc group test, the test returns a usable value.
+     *
+     * If there is no caching and the second person object only exists for the recursive call, then the implementation
+     * works.
+     *
+     * Also, if the person object is cached and used twice, then the group key with the ad hoc tester is not added to
+     * the containing group keys during the recursion but is added (or not) after the test call returns positive.
+     */
+    @Override
+    public boolean test(IPerson person) {
+        String personHash = person.getEntityIdentifier().getKey() + groupsHash + Thread.currentThread().getId();
+        logger.debug("Entering test() for {}", personHash);
+        if (currentTests.getQuiet(personHash) != null) {
+            logger.warn("Returning from test() for {} due to recusion for person = {}", personHash, person.toString());
```

Comment:
Should this be a warning or info?  Does this mean the groups were setup recursively (user error which warrants a warning and more information to explain how the recursion occurred) or that is a natural aspect of this group tester having include groups (an info or debug because it is not a user error)?